BraintreeModels/src/main/java/com/braintreepayments/api/PaymentMethodNonce.java

```diff
@@ -1,186 +1,29 @@
 
     /**
      * @return The description of this PaymentMethod for displaying to a customer, e.g. 'Visa ending in...'
      */
-    public String getDescription() {
-        return mDescription;
-    }
+    String getDescription();
 
     /**
      * @return {@code true} if this payment method is the default for the current customer, {@code false} otherwise
      */
-    public boolean isDefault() {
-        return mDefault;
-    }
+    boolean isDefault();
 
     /**
      * @return The type of this PaymentMethod for displaying to a customer, e.g. 'Visa'. Can be used
      *          for displaying appropriate logos, etc.
      */
-    public abstract String getTypeLabel();
-
-    /**
-     * Parses a response from the Braintree gateway for a list of payment method nonces.
-     *
-     * @param jsonBody Json-formatted String containing a list of {@link PaymentMethodNonce}s
-     * @return List of {@link PaymentMethodNonce}s contained in jsonBody
-     * @throws JSONException if parsing fails
-     */
-    static List<PaymentMethodNonce> parsePaymentMethodNonces(String jsonBody)
-            throws JSONException {
-        JSONArray paymentMethods = new JSONObject(jsonBody).getJSONArray(
-                PAYMENT_METHOD_NONCE_COLLECTION_KEY);
-
-        if (paymentMethods == null) {
-            return Collections.emptyList();
-        }
-
-        List<PaymentMethodNonce> paymentMethodsNonces = new ArrayList<>();
-        JSONObject json;
-        PaymentMethodNonce paymentMethodNonce;
-        for(int i = 0; i < paymentMethods.length(); i++) {
-            json = paymentMethods.getJSONObject(i);
-            paymentMethodNonce = parsePaymentMethodNonces(json,
-                    json.getString(PAYMENT_METHOD_TYPE_KEY));
-            if (paymentMethodNonce != null) {
-                paymentMethodsNonces.add(paymentMethodNonce);
-            }
-        }
-
-        return paymentMethodsNonces;
-    }
-
-    /**
-     * Parses a {@link PaymentMethodNonce} from json.
-     *
-     * @param json {@link String} representation of a {@link PaymentMethodNonce}.
-     * @param type The {@link String} type of the {@link PaymentMethodNonce}.
-     * @return {@link PaymentMethodNonce}
-     * @throws JSONException if parsing fails
-     */
-    @Nullable
-    static PaymentMethodNonce parsePaymentMethodNonces(String json, String type) throws JSONException {
-        return parsePaymentMethodNonces(new JSONObject(json), type);
-    }
-
-    /**
-     * Parses a {@link PaymentMethodNonce} from json.
-     *
-     * @param json {@link JSONObject} representation of a {@link PaymentMethodNonce}.
-     * @param type The {@link String} type of the {@link PaymentMethodNonce}.
-     * @return {@link PaymentMethodNonce}
-     * @throws JSONException if parsing fails
-     */
-    @Nullable
-    static PaymentMethodNonce parsePaymentMethodNonces(JSONObject json, String type) throws JSONException {
-        switch (type) {
-            case CardNonce.TYPE:
```

Comment:
I'm a little concerned that by removing this logic and giving merchants just generic `BraintreeNonce` objects, there will be some info merchants will no longer be able to access on vaulted payment methods. I'm not 100% sure how much of a concern this is since it depends on what merchants are doing with those vaulted payment methods. But it might be safer to make sure that they can access all of the same information as before.

I know we want to separate out the specific nonce-types into their respective modules, so that they don't all live in Core. Would it be possible to do a run-time check for those classes in Core so that we can still parse the vaulted payment method response into the specific nonce types? We'd still need to return the base class to the merchant, but they'd be able to cast to the more specific type of nonce if needed, and all of the relevant info would be there.

Reply:
Yeah I like that idea. Right now we have [copy constructors](https://github.com/braintree/braintree_android/blob/8825a4d50ded576dd7fcefc1962bc61e600d25e2/GooglePay/src/main/java/com/braintreepayments/api/GooglePayCardNonce.java#L44) for the merchant to do a cast-like operation to access nonce-type specific information. We could leverage `compileOnly` here to take the burden off of the merchant, then they can use `instanceOf` for introspection.